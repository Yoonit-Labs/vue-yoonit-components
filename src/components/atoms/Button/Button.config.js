module.exports = {
  variation: {
    options: [
      'base',
      'outline',
      'clear'
    ]
  },
  buttonSize: {
    options: [
      'normal',
      'short',
      'xs',
      'large'
    ]
  },
  fill: {
    options: [
      'primary',
      'danger',
      'light'
    ]
  },
  iconPosition: {
    options: [
      'left',
      'right'
    ]
  },
  iconSize: {
    options: [
      'large',
      'normal',
      'small'
    ]
  },
  iconStyle: {
    options: [
      'solid',
      'regular'
    ]
  },
  animation: {
    options: [
      'none',
      'pulse'
    ]
  }
}
