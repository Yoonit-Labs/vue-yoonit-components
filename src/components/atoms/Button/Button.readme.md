# Yonnit Components

[**Home**](https://cyberlabs.ai/) | [**Atoms**](https://cyberlabs.ai/) | [**Molecules**](https://cyberlabs.ai/) | [**Bosons**](https://cyberlabs.ai/) | [**Quarks**](https://cyberlabs.ai/)

## Button

The component is used to illustrate a user and displays the profile picture or the beginning of the word

## Usage

To use the component, simply invoke it in your template with the desired props

### Input
`App.vue`
```vue
<template>
  <yoo-button
    variation="base"
    buttonSize="normal"
    fill="primary"
    animation="none"
    text="Button Base"
  />
  
  <yoo-button
    variation="outline"
    buttonSize="normal"
    fill="primary"
    animation="none"
    text="Button Outline"
  />
  
  <yoo-button
    variation="base"
    buttonSize="normal"
    fill="danger"
    icon="check"
    iconPosition="left"
    iconSize="normal"
    iconStyle="solid"
    animation="none"
    text="Button danger + icon check"
  />
</template>
```
### Output

The output of the above code will be:

<img src="https://github.com/Yoonit-Labs/vue-yoonit-components/blob/development/public/readme-img/button-example.png" width="400">

### Props

| Name               | Input/Format                                  | Description                                                                 |
| -                  | -                                             | -                                                                           |
| variation          | base, outline, clear                          | Sets variation for the button - Base (filled) / Outline (bordered, colorless) / Clear (Borderless and colorless)|
| buttonSize         | normal, short, xs                             | Sets size for the button                                                    |
| fill               | primary, danger, light, dark                  | Sets color for the button                                                   |
| iconPosition       | left, right                                   | Defines the position of the icon in relation to the button text             |
| iconSize           | small, normal, large                          | When there is an icon, set the icon size                                    |
| iconStyle          | regular, solid                                | Sets the icon type to filled or bordered only                               |
| animation          | none, pulse                                   | Set a pulse animation to the button                                         |
| iconCustomColor    | Hexadecimal color                             | Sets custom color for the icon                                              |
| text               | String                                        | Set text for the button                                                     |
| icon               | String                                        | Sets an icon on the button based on fontawesome                             |
| disabled           | true/false                                    | Disables the button                                                         |
| hover              | true/false                                    | Enables the style change on the button hover                                |
| active             | true/false                                    | Enables the style change in the button active                               |

## To contribute and make it better

Clone the repo, change what you want and send PR.

Contributions are always welcome!

---

Code with ‚ù§ by the [**Cyberlabs AI**](https://cyberlabs.ai/) Front-End Team
