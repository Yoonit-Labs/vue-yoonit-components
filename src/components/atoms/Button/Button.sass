@import '~@yoonit/style'
@import '~@yoonit/style/atoms/YooButton'