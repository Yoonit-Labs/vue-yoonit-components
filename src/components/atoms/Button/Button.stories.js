import yooButton from './Button.vue'

export default {
  title: 'Yoo/Button',
  component: yooButton,
  argTypes: {
    variation: {
      control: {
        type: 'select',
        options: [
          'base',
          'outline',
          'clear'
        ]
      }
    }
  }
}

const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { yooButton },
  template:
    '<yoo-button v-bind="$props"></yoo-button>'
})

export const Default = Template.bind({})
Default.args = { text: "Yoo Button" }
