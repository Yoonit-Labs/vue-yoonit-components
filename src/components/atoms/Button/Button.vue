<template lang="pug">
.yoo__container--button
  button(
    :class="['yoo-btn', ...takeModifier]"
    :disabled="disabled"
    @click="doClick"
  )
    i(
      v-if="icon.length && iconPosition === 'left'"
      :class="['yoo-btn__icon', takeIconStyle, takeIconName, ...takeModifierIcon]"
      :style="takeIconColor"
    )
    p.yoo-btn__text {{takeText}}

    i(
      v-if="icon.length && iconPosition === 'right'"
      :class="['yoo-btn__icon', takeIconStyle, takeIconName, ...takeModifierIcon]"
      :style="takeIconColor"
    )
</template>
<script>

export default {
  name: 'Button',
  props: {
    variation: {
      type: String,
      default: 'base',
      required: false,
      validator: value =>
        [
          'base',
          'outline',
          'clear'
        ]
          .includes(value)
    },
    buttonSize: {
      type: String,
      default: 'normal',
      required: false,
      validator: value =>
        [
          'normal',
          'short',
          'large'
        ]
          .includes(value)
    },
    fill: {
      type: String,
      default: 'primary',
      required: false,
      validator: value =>
        [
          'primary',
          'danger'
        ]
          .includes(value)
    },
    icon: {
      type: String,
      required: false,
      default: ''
    },
    iconPosition: {
      type: String,
      required: false,
      default: 'left',
      validator: value =>
        [
          'left',
          'right'
        ]
          .includes(value)
    },
    iconSize: {
      type: String,
      required: false,
      default: 'normal',
      validator: value =>
        [
          'large',
          'normal',
          'small'
        ]
          .includes(value)
    },
    iconStyle: {
      type: String,
      required: false,
      default: 'solid',
      validator: value =>
        [
          'solid',
          'regular'
        ]
          .includes(value)
    },
    animation: {
      type: String,
      required: false,
      default: 'none',
      validator: value =>
        [
          'none',
          'pulse'
        ]
          .includes(value)
    },
    disabled: {
      type: Boolean,
      required: false,
      default: false
    },
    text: {
      type: String,
      required: false,
      default: ''
    },
    hover: {
      type: Boolean,
      required: false,
      default: false
    },
    active: {
      type: Boolean,
      required: false,
      default: false
    },
    iconCustomColor: {
      type: String,
      default: '',
      validator (value) {
        if (value.includes('#')) {
          const rx = /^#([0-9a-f]{6}|[0-9a-f]{3})$/i
          return value.match(rx)
        } else {
          return value === ''
        }
      }
    }
  },
  data: () => ({}),
  created () {},
  mounted () {},
  updated () {},
  beforeDestroy () {},
  destroyed () {},
  components: {},
  computed: {
    takeText () {
      return this.text
    },
    takeIconColor () {
      return {
        color: `${this.iconCustomColor}`
      }
    },
    takeHover () {
      if (this.hover) {
        return 'hover'
      }
      return ''
    },
    takeActive () {
      if (this.active) {
        return 'active'
      }
      return ''
    },
    takeDisabled () {
      if (this.disabled) {
        return 'disabled'
      }
      return ''
    },
    takeModifierIcon () {
      const block = 'yoo-btn__icon'
      const classList = []
      if (this.icon) {
        if (this.iconSize) {
          classList
            .push(
              `${block}--${this.iconSize}`
            )
        }
        if (this.iconPosition && this.text) {
          classList
            .push(
              `${block}--${this.iconPosition}`
            )
        }
      }
      return classList
    },
    takeModifier () {
      const block = 'yoo-btn'
      const classList = []

      if (this.fill) {
        classList
          .push(
            `${block}__fill--${this.fill}`
          )
      }
      if (this.fill && this.hover && !this.disabled) {
        classList
          .push(
            `${block}--${this.takeHover}`
          )
      }
      if (this.fill && this.active && !this.disabled) {
        classList
          .push(
            `${block}--${this.takeActive}`
          )
      }
      if (this.fill && this.disabled) {
        classList
          .push(
            `${block}--${this.takeDisabled}`
          )
      }
      if (this.variation && (this.variation !== 'base')) {
        classList
          .push(
            `${block}--${this.variation}`
          )
      }

      if (this.format && (this.format !== 'block')) {
        classList
          .push(
            `${block}--${this.format}`
          )
      }

      if (this.size && (this.size !== 'normal')) {
        classList
          .push(
            `${block}--${this.size}`
          )
      }

      if (this.buttonSize && (this.buttonSize !== 'normal')) {
        classList
          .push(
            `${block}--${this.buttonSize}`
          )
      }

      if (this.animation && (this.animation !== 'none')) {
        classList
          .push(
            `${block}--animation-${this.animation}`
          )
      }
      return classList
    },
    takeIconName () {
      return `fa-${this.icon}`
    },
    takeIconStyle () {
      return this.iconStyle === 'solid'
        ? 'fas' // fontawesome solid
        : 'far' // fontawesome regular
    }
  },
  methods: {
    doClick (event) {
      if (event) {
        this.$emit('onClick', event)
      }
    }
  },
  filters: {},
  watch: {}
}
</script>

<style src="./Button.sass" lang="sass" scoped></style>
