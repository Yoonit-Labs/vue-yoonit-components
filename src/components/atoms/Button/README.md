# Yoonit Components

[Home](https://github.com/Yoonit-Labs/vue-yoonit-components/blob/development/README.md) | [Atoms](https://github.com/Yoonit-Labs/vue-yoonit-components/blob/development/README.md#atoms) | [Molecules](https://github.com/Yoonit-Labs/vue-yoonit-components/blob/development/README.md#molecules) | [Bosons](https://github.com/Yoonit-Labs/vue-yoonit-components/blob/development/README.md#bosons) | [Quarks](https://github.com/Yoonit-Labs/vue-yoonit-components/blob/development/README.md#quarks)

## YooButton

The component is used to illustrate a user and displays the profile picture or the beginning of the word

## Usage

To use the component, simply invoke it in your template with the desired props

### Input
`App.vue`
```vue
<template>
  <yoo-button
    variation="base"
    buttonSize="normal"
    fill="primary"
    animation="none"
    text="Button Base"
  />
  
  <yoo-button
    variation="outline"
    buttonSize="normal"
    fill="primary"
    animation="none"
    text="Button Outline"
  />
  
  <yoo-button
    variation="base"
    buttonSize="normal"
    fill="danger"
    icon="check"
    iconPosition="left"
    iconSize="normal"
    iconStyle="solid"
    animation="none"
    text="Button danger + icon check"
  />
</template>
```
### Output

The output of the above code will be:

<img src="../../../../public/readme-img/button.png" width="400">

### Props

| Parameter          | Type    | Default | Valid values                              | Description                                    | Required |
|--------------------|---------|---------|-------------------------------------------|------------------------------------------------|----------|
| `variation`        | String  | base    |<ul><li>base</li><li>outline</li><li>clear</li>     | Sets variation for the button - Base (filled) / Outline (bordered, colorless) / Clear (Borderless and colorless)|
| `fill`             | String  | primary |<ul><li>primary</li><li>danger</li><li>light</li><li>dark</li></ul>| Sets color for the button   |
| `buttonSize`       | String  | normal  |<ul><li>normal</li><li>short</li><li>xs</li></ul>  | Sets size for the button    |
| `text`             | String  |   ''    | String                                        | Set text for the button     |
| `icon`             | String  |   ''    | String                                        | Sets an icon on the button based on fontawesome    |
| `iconPosition`     | String  | left    | <ul><li>left</li><li>right</li></ul>          | Defines the position of the icon in relation to the button text    |
| `iconSize`         | String  | normal  | <ul><li>small</li><li>normal</li><li>large</li></ul> | When there is an icon, set the icon size      |
| `iconStyle`        | String  | solid   | <ul><li>regular</li><li>solid</li></ul>              | Sets the icon type to filled or bordered only |
| `iconCustomColor`  | String  |   ''    | Any color i.e #00000                             | Sets custom color for the icon       |
| `animation`        | String  | none    | <ul><li>none</li><li>pulse</li></ul>          | Set a pulse animation to the button        |
| `disabled`         | Boolean | false   | true or false                                 | Disables the button       |
| `hover`            | Boolean | false   | true or false                                 | Enables the style change on the button hover    |
| `active`           | Boolean | false   | true or false                                 | Enables the style change in the button active  |

#

 #### [**Next component**](../CheckButton/README.md) :arrow_forward:

 #### :arrow_backward: [**Previous component**](../Avatar/README.md)

#

## To contribute and make it better

Clone the repo, change what you want and send PR.

Contributions are always welcome!

---

Code with ‚ù§ by the [**Cyberlabs AI**](https://cyberlabs.ai/) Front-End Team
