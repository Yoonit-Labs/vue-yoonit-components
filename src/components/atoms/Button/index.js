import yooButton from './Button.vue'
export default yooButton
