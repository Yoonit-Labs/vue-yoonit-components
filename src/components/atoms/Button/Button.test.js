import { mount } from '@vue/test-utils'
import Button from '@/components/atoms/Button/Button.vue'

const classBlock = 'yoo-btn'
const SlotText = 'Default Slot Text'
const ButtonProps = {
  variation: ['base', 'outline', 'clear'],
  buttonSize: ['normal', 'short', 'large'],
  fill: ['primary', 'danger'],
  iconPosition: ['left', 'right'],
  iconSize: ['large', 'normal', 'small'],
  iconStyle: ['solid', 'regular'],
  animation: ['none', 'pulse']
}

const mountButton = () => {
  return mount(Button, {
    slots: { default: SlotText }
  })
}

describe('Button Component', () => {
  let wrapper
  beforeEach(() => {
    wrapper = mountButton()
    // console.log('wrapper', wrapper)
  })

  it('Matches Snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('Loads the Component HTML', () => {
    expect(wrapper.classes('yoo__container--button')).toBe(true)
    expect(wrapper.find('button.yoo-btn').exists()).toBe(true)
  })

  describe('Props', () => {

    describe('variation', () => {
      ButtonProps.variation.forEach(variation => {
        it(`Accept prop variation: ${variation}`, async () => {
          await wrapper.setProps({ variation })
          if (variation !== 'base') {
            expect(wrapper.find(`.${classBlock}--${variation}`).exists()).toBe(true)
          }
        })
      })
    })

    describe('buttonSize', () => {
      ButtonProps.buttonSize.forEach(buttonSize => {
        it(`Accept prop buttonSize: ${buttonSize}`, async () => {
          await wrapper.setProps({ buttonSize })
          if (buttonSize !== 'normal') {
            expect(wrapper.find(`.${classBlock}--${buttonSize}`).exists()).toBe(true)
          }
        })
      })
    })

    describe('fill', () => {
      ButtonProps.fill.forEach(fill => {
        it(`Accept prop fill: ${fill}`, async () => {
          await wrapper.setProps({ fill })
          expect(wrapper.find(`.${classBlock}__fill--${fill}`).exists()).toBe(true)
        })
      })
    })

    describe('iconPosition', () => {
      ButtonProps.iconPosition.forEach(iconPosition => {
        it(`Accept prop iconPosition: ${iconPosition}`, async () => {
          await wrapper.setProps({ iconPosition })

          expect(wrapper.find(`.${classBlock}__icon--${iconPosition}`).exists()).toBe(false)

          await wrapper.setProps({ text: 'Custom Text' })

          expect(wrapper.find(`.${classBlock}__icon--${iconPosition}`).exists()).toBe(true)
        })
      })
    })

    describe('iconSize', () => {
      ButtonProps.iconSize.forEach(iconSize => {
        it(`Accept prop iconSize: ${iconSize}`, async () => {
          await wrapper.setProps({ iconSize })
          expect(wrapper.find(`.${classBlock}--${iconSize}`).exists()).toBe(false)
        })
      })
    })

    //   ButtonProps.iconStyle.forEach(iconStyle => {
    //     it(`Accept prop iconStyle: ${iconStyle}`, async () => {
    //       await wrapper.setProps({ iconStyle })
    //       // expect(wrapper.find(`.${SubtitleClass}--${size}`))
    //     })
    //   })
    //   ButtonProps.animation.forEach(animation => {
    //     it(`Accept prop animation: ${animation}`, async () => {
    //       await wrapper.setProps({ animation })
    //       // expect(wrapper.find(`.${SubtitleClass}--${size}`))
    //     })
    //   })
  })
})
