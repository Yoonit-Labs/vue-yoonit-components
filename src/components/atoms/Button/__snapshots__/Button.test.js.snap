// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component Matches Snapshot 1`] = `
<div class="yoo__container--button"><button class="yoo-btn yoo-btn__fill--primary">
    <!---->
    <p class="yoo-btn__text"></p>
    <!---->
  </button></div>
`;
